js/ui/appDisplay.js

```diff
@@ -282,6 +282,20 @@ var BaseAppView = new Lang.Class({
             // The icon image changed
             if (newIconInfo && !newIconInfo.equal(oldIconInfo))
                 return true;
+
+            if (item instanceof Shell.App) {
```

Comment:
This is … not pleasant, right? One could put the comparison function onto the `ShellApp` but then you still need a typecheck to discover whether we have one of those or a ShellDesktopDirInfo. Or we could construct the new ViewIcon (probably an expensive operation for what's meant to be a reasonably quick check…), and add the comparison function to its two subclasses.

Reply:
I think this approach is fine for now.

Reply:
But not in the long run, of course.